O365 phishlet

JamesCullum opened the pull request:
Tested and catches all logins from ADFS and O365. Depending on the local ADFS implementation this might need to be extended with community feedback.

In addition, there is a workaround for the missing detection of the :443. Fields on the sub_filters & proxy_hosts need to be manually filled into the yaml before launch. By allowing to use params for those things (separate PR) this can be avoided in the future.

kgretzky commented:
Thanks a lot for your contribution!

Can anyone with access to O365 account, check if it works?

kgretzky commented:
Right. That `:443` suffix you mentioned, was PRed before, but I haven't implemented it yet #122

gettheguru commented:
Hi, I gave it a try but the first thing I noticed was a certificate error;

Lure URL: 
https://login.<mybasedomain>/<lureID>

But on visit there is a cert warning:
Your connection is not private
Attackers might be trying to steal your information from www.<mybasedomain> (for example, passwords, messages or credit cards). Learn more
NET::ERR_CERT_COMMON_NAME_INVALID

This server could not prove that it is www.<mybasedomain>; its security certificate is from login.<mybasedomain>. This may be caused by a misconfiguration or an attacker intercepting your connection.

Proceed to www.<mybasedomain> (unsafe)

I haven't had an issue with the letsencrypt certificates before in phishlets - so not sure if this is specific to this phishlet or not

gettheguru commented:
> I haven't had an issue with the letsencrypt certificates before in phishlets - so not sure if this is specific to this phishlet or not

The issue was on my end - the phishlet appears to work fine, although I haven't been able to make use of the captured tokens yet after importing them to recreate a captured session. I am unable to confirm ADFS though as that is not in use in my organisation.

JamesCullum commented:
This usually happens because you are re-using a hostname. Use `rm -rf /root/.evilginx/crt/<hostname>` to delete it and restart evilnginx. It should then re-generate the certificates.

> I need to digest this idea
Considering that the previous one is still open for quite some time, can you give a rough timeline?

pawp81 commented:
Great work! Does this phishlet work with Office 365 MFA-enabled account? 
I installed phishlet successfully but after importing cookie to Chrome I am getting prompt for second-factor (no prompt for login name or password). MFA is activated on the account by Azure AD.
When account is not MFA-enabled I can get full way in with ESTSAUTH and ESTSAUTHPERSISTENT cookies.

pawp81 commented:
I think the reason might be that first pair of ESTSAUTH + ESTSAUTHPERSISTENT cookies is issued after username+password auth (https://login.microsoftonline.com/common/login). Then new pair of ESTSAUTH + ESTSAUTHPERSISTENT cookies with new values is issued after second factor auth https://login.microsoftonline.com/common/SAS/ProcessEnd). However evilginx only captures first pair of cookies. It confirmed that with "all authorization tokens intercepted!" despite the fact that I still had to perform 2nd factor auth.
@kgretzky can you confirm that evilginx is not greedy? If it has all the cookies defined in the phishlet to be capture it will not overwrite them with new cookies that arrive after 2nd factor authentication?

gettheguru commented:
Just commenting to echo I have the experience with the authentication cookies as @pawp81. It would be great to hear from @kgretzky if there is a means to overwrite this behaviour in the phishlet or any other way?

pawp81 commented:
OK I made it work. I added to @JamesCullum  phishlet following code:

`auth_urls:
  - '/kmsi'` 
Thanks to that evilginx will capture cookies until user is redirected to login.microsoftonline.com/kmsi which happens after 2nd factor authentication. It will save last ESTSAUTH cookie. 
Btw I am not sure if ESTSAUTHPERSISTENT cookie needs to be captured at all. I was importing only ESTSAUTH cookied and it worked fine.

gettheguru commented:
Can confirm that @pawp81 MFA change works for me, using all the MFA auth types I tried (SMS, call, authenticator app) - I'm able to capture and recreate the sessions.

However I've noticed that this does break my redirect_url on the lures, whereas before it would redirect after authentication, it no longer is able to, unless I re-visit the initial lure URL for a second time. Any idea how to get this functionality back while keeping the MFA token capture??

JamesCullum commented:
Would it be possible for you to provide me with access to such a MFA login (for a dummy/debugging account)? Then I can support you in this. As my organization doesn't utilize MFA, I can't test it myself.

pawp81 commented:
How can I communicate with you privately?

JamesCullum commented:
You can send me an email to nicolas.giese@web.de (will edit this comment later to remove).

JamesCullum commented:
Thanks to @pawp81 I was able to test it for 2FA and was able to fix it, so that the `redirect_url` works and cookies always get captured. This works for me independantly if the user remembers log ins or not.

Compatibility with ADFS is tested and still works, so it shouldn't break anything.

@pawp81 @gettheguru 
Can both of you test the newest version and verify whether or not it works?

pawp81 commented:
@JamesCullum  I tested and new phishlet works fine with 2FA and without 2FA. I have a small problem though with federated account authenticating via ADFS. There is no message that all tokens have been intercepted. However the session with cookies, username and password is being saved.

root2019x commented:
> 
> 
> Thanks to @pawp81 I was able to test it for 2FA and was able to fix it, so that the `redirect_url` works and cookies always get captured. This works for me independantly if the user remembers log ins or not.
> 
> Compatibility with ADFS is tested and still works, so it shouldn't break anything.
> 
> @pawp81 @gettheguru
> Can both of you test the newest version and verify whether or not it works?

Mind sharing it here also? rootvpstest2019@outlook.com . Will test with ADFS.

gettheguru commented:
Hi @JamesCullum - all working for me except the redirect_url only works on 2nd visit to the lure address. I do see 'all authorization tokens intercepted!' in the console though. You mentioned that the redirect_url works now for you - any idea what I'm doing wrong? FYI there is no overlap between the phishing URL and the redirect URL

JamesCullum commented:
Hey @gettheguru,

With the new method it redirects immediately for me.
Do you have a lure configuration for me I can try?

@pawp81 Is redirect_url working for you?

gettheguru commented:
Hey @JamesCullum, thanks for the swift reply.

Pretty simple really;
lures create o365
lures edit redirect_url 0 https://www.google.com

Second visit takes me to google as expected. First visit holds me on the office.com proxied via my phishing site (www.xxxxx/?auth=2) with no re-direct. I've tried both selecting "no" and "yes" on the remember my details page.

JamesCullum commented:
Alright, I took a closer look into this now. The issue appears to be in the main code itself - it triggers the redirect at the wrong time.

O365 redirects the user to the final www.office.com landing page with a JS form press. However, the browser (for me, FF) triggers a favicon page. Before, this is always a 404, that's why it retries it. That page is redirected to `redirect_url` but the submitted page isn't.

![console image](https://i.imgur.com/fofTgVA.png)

Theoretically, this would require code changes on the intercept. As the other PRs look like @kgretzky is not a big fan of changing the code base, this is out of scope for now.

It was also not possible to use the new javascript injection possibilities. Due to missing debugging capabilities, it is not easy to find out why.

However, to still be able to use just the plugin to use the desired redirect, I added a rewrite of the final submission. You can see in the newest commit that I added a small manual for that. Basically, uncomment the line for the redirect and put the desired URL.

@gettheguru Can you test and confirm whether or not this works for you as well, please?

gettheguru commented:
Hi @JamesCullum,

Thank you very much for your efforts on this, it is much appreciated. I also found I was unable to implement the javascript injection for whatever reason.

I can confirm your workaround is working perfectly for me. And everything else appears fully functional that I am able to test.

Given how prevalent o365 is at the moment I think this would make a great addition to the evilginx2 base branch.

Thank you again!

kgretzky commented:
Ok, so as I understand the redirect triggers for the `favicon.ico` only and later it doesn't redirect?
I've changed the redirection to trigger only once in latest version, so this may be causing the issue.

I think the reasonable way to counter it would be to set a maximum number of redirects to trigger for each session (currently default is `1`) and only trigger redirects for `text/html` Content-Type.

JamesCullum commented:
Hey @kgretzky, that appears to be the issue. I think the main part is the filtering for the content type, which should be easy to imlement.

Do you want to merge the PR then? It probably makes sense to track the bug in a different issue.

kgretzky commented:
Thanks. I will try to fix it over the weekend in the code and then the phishlet should work fine in its original form, so I don't think a merge now is needed.

Thanks for the screenshot of the requests history - it made it clear to me where the issue resides.

Will let you know here.

kgretzky commented:
@JamesCullum I've just pushed the redirection fix. Can you pull and tell me if you still have an issue with redirection not triggering for proper request?

JamesCullum commented:
I need to build with the latest master, right? Or is there a compiled version?

JamesCullum commented:
Alright, I build it from source and tested it.
I now get redirected without having to activate the workaround.
I will therefore remove that part from the phishlet.

![success screenshot](https://i.imgur.com/r60keGK.png)

kgretzky commented:
Great job @JamesCullum 
Thank you for your contribution! Merging.

alimukarramm commented:
> Many thks, work fine with ADFS

hey frdaz. sorry for being ignorant but I wonder if you please let me know how that ADFS worked for you. I have a test organization with ADFS configured. I used the O365 phishlet and followed all the necessary steps. I got the O365 page even but upon redirecting to my Organization login page it says that I am not authorized to use this location (something like that). If possible could you please drop few lines how you achieved this. 

Thanks in advance.

JamesCullum commented:
Did you put the correct ADFS url into the config file? Maybe show us your config, so that we can help.

alimukarramm commented:
Hi James, thanks for your swift reply. Sure, will do once I login to my droplet. Thanks again.

alimukarramm commented:
Hi, I just checked the o365.yaml file but couldn't find a way to set my ADFS URL. Neither, I could find the redirect_url option. Please advice and I apologize again for my ignorance. Thanks.

JamesCullum commented:
There are placeholders in the file that you need to replace. If it doesn't work, you probably didn't do that. There are instructions in the file to support you.

alimukarramm commented:
Thanks a lot for your speedy replies and help...

frdaz commented:
Hello, i used ADFS, and i modified the yaml placeholders by my company but nothing work, in login page cannot find my username@federated domain. i search if i find anything.

Yulia1991lab commented:
@pawp81 please share your email address.

Thanks

Yulia1991lab commented:
@pawp81  email me once at yulia19@uymail.com.

Big Thanks

Amrougaza commented:
Hi,
i have used this O365 [phishlet] 
[O365 phishlet.txt](https://github.com/kgretzky/evilginx2/files/3898085/O365.phishlet.txt)
and i have changed the ADFS values corresponding to my company, but it doesn't work for me.
   id  | phishlet  |         username          |    password      |  tokens   |  remote ip    |       time        |
+-----+-----------+---------------------------+------------------+-----------+---------------+------------
| 17  | o365      |                                    |                          | none      |                 | 2019-11-27 17:21  |

Thank you for help.

YusufHegazy commented:
Hi!
I'm trying to use this phishlet but on typing my email it redirects me to a https://login.live.com/ page displaying "invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application."
does this have something to do with the phishlet, or it's on my side?

gettheguru commented:
That’ll mean it’s not an office 365 account, it’s more like a Hotmail or Live account.

On 17 Feb 2020, at 21:51, YusufHegazy <notifications@github.com> wrote:
> 
> ﻿
> Hi!
> I'm trying to use this phishlet but on typing my email it redirects me to a https://login.live.com/ page displaying "invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application."
> does this have something to do with the phishlet, or it's on my side?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

YusufHegazy commented:
> office 365

oh my gosh, how dumb can i be.
thanks you!

0mikron commented:
Hey @JamesCullum 

There seems to be an issue when I try to use it on my own ADFS.

I've commented out the lines on the o365.yaml like this:

~~~yaml
- {phish_sub: 'adfs', orig_sub: 'adfs', domain: 'originadomain.com', session: true, is_landing:false}
- {phish_sub: 'adfs', orig_sub: 'adfs', domain: 'originaldomain.com:443', session: true, is_landing:false}
.
.
- {triggers_on: 'adfs.originaldomain.com', orig_sub: 'login', domain: 'microsoftonline.com', search: 'https://{hostname}', replace: 'https://{hostname}', mimes: ['text/html', 'application/json', 'application/javascript']} 
~~~

But when I enter my credentials I get the following error:

`The requested federation realm object ;http://adfs.phishdomain.com:443/adfs/services/trust; does not exist.`

When I try to remove the `:443` from the request I get redirected properly and asked for 2FA, but you will need to be in the client's machine to do it (I did it with match/replace in Burp while proxying the traffice from Evilginx2)

Is there a way to add a trigger that will do it for me?

Thanks 😄

JamesCullum commented:
`but you will need to be in the client's machine to do it` what do you mean with that?

0mikron commented:
Yea that sentence does not really make sense..

One hacky way of solving the issue was to proxy the Evilginx server through Burp and add a match/replace rule which replaces every `:443/adfs/services/trust` to `/adfs/services/trust` it finds. 

I tried to write a `sub_filter` but failed horribly 😅.

My final solution was to edit the `http_proxy.go` file by adding some code which searches for the string `:443/adfs` in the `req.Body` for the specific url path, in my case `/login.srf`  and remove the `:443`. 

This worked great, however I'm pretty sure that this should be possible with a `sub_filter`.

I can make a pull request so that you have a more clear of idea of my workaround.

Cheers 😄

JamesCullum commented:
The bug with the :443 is known and was reported already. Not sure if it has been fixed, but thats why the sub_filter is there - in most cases it works well, but its impossible to tell from here why it doesnt for you.

0mikron commented:
I think the issue with the suggested `sub_filter` is the is that the request is for `http://originaldomain.com:443` and not `https`.

Shouldn't an extra trigger like this catch the above? 

`  - {triggers_on: 'adfs.originalsubdomain.com', orig_sub: 'login', domain: 'microsoftonline.com', search: 'http://{hostname}', replace: 'http://{hostname}', mimes: ['text/html', 'application/json', 'application/javascript']}
`

gettheguru commented:
Are there any gotchas or special processes when taking the authorisation cookies with office 365? I did a quick test this evening but wasn’t able to make use of the stolen cookies (imported with editthiscookie) but seemingly unable to get login.microsoftonline.com to pay attention to them

pretech86 commented:
i get an ssl certificate and put it in ./evilginx/crt/ and it working well , and when i run apache2 and check the certificate it run correctly with no error and the browser trust my domain , but when i stop apache2 and run evilginx the browser still not trust the certificate and the facebook page not like the real one , i already catch the token but it give me 

WARN: Cannot handshake client static.xx.fbcdn.net remote error: tls: bad certificate 

please any one can help me sad.noscom@yahoo.com

d3v3lop3rtuta commented:
> Tested and catches all logins from ADFS and O365. Depending on the local ADFS implementation this might need to be extended with community feedback.
> 
> In addition, there is a workaround for the missing detection of the :443. Fields on the sub_filters & proxy_hosts need to be manually filled into the yaml before launch. By allowing to use params for those things (separate PR) this can be avoided in the future.

is there a workaround to make ADFS autodetect domain instead of having to manually specify domain? I'm think maybe a regex implementation? Thank you .

Thurnun commented:
Hello, has anyone else tested Evilginx2 with O365 using MFA recently? If so is it still working for you?

I would like to confirm whether the issue is on my end or if something has changed in O365 to prevent this.

I am having issues similar to what is mentioned earlier in this pull request. My phishing page will capture username/password and session tokens before MFA prompts. When I enter the username/pass (captured), my phishing page is directed to office.com (www.xxxxx/?auth=2). Here I can only click on Outlook and/or apps(as the signed in user), once I click on Outlook I leave my phishing page and am prompted to put in my MFA (authenicator app) back at login.microsoftonline.com but not on my phishing page. 

I have tried using auth_urls in the o365.yaml, specifying '/kmsi' as mentioned earlier. Though this does not work because "/kmsi" happens before I am prompted for MFA.

I am running Evilginx2 latest, built from source.

h4k4zworld commented:
> > Tested and catches all logins from ADFS and O365. Depending on the local ADFS implementation this might need to be extended with community feedback.
> > In addition, there is a workaround for the missing detection of the :443. Fields on the sub_filters & proxy_hosts need to be manually filled into the yaml before launch. By allowing to use params for those things (separate PR) this can be avoided in the future.
> 
> is there a workaround to make ADFS autodetect domain instead of having to manually specify domain? I'm thinking maybe a regex implementation? Thank you .

You should cc @JamesCullum

JamesCullum commented:
As long as @kgretzky is not merging changes outside of phishlets, I don't believe it is possible. I'm also not sure I understand your organizational setup, which would allow you to phish several ADFS domains. If it is different organizations and you are a professional pentesting company, you should separate the data anyways and put it on different servers.

donjuanme commented:
> That’ll mean it’s not an office 365 account, it’s more like a Hotmail or Live account.
> […](#)
> On 17 Feb 2020, at 21:51, YusufHegazy ***@***.***> wrote: ﻿ Hi! I'm trying to use this phishlet but on typing my email it redirects me to a https://login.live.com/ page displaying "invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application." does this have something to do with the phishlet, or it's on my side? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

is there any hotmail or live account yaml available ?

gettheguru commented:
Yep the outlook.yaml phislet

> On 11 Jul 2020, at 16:07, donjuanme <notifications@github.com> wrote:
> 
> ﻿
> That’ll mean it’s not an office 365 account, it’s more like a Hotmail or Live account.
> …
> On 17 Feb 2020, at 21:51, YusufHegazy @.***> wrote: ﻿ Hi! I'm trying to use this phishlet but on typing my email it redirects me to a https://login.live.com/ page displaying "invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application." does this have something to do with the phishlet, or it's on my side? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> is there any hotmail or live account yaml available ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

hichemhero commented:
greetings, 
i have found a problem in the current Facebook phishlet, it only records the password on the first attempt only
say per example the victim has entered a wrong pass,  the wrong pass will be recorded as the first attempt, but after the user inserts the correct one or other attempts it doesn't record it on #sessions, however the username is being recorded each time and works fine.
can somebody check if its an actual bug or its just happening with me? 
thanks

grubbychicken commented:
Hi @JamesCullum,

I seem to be having the same `http......:443` issue. 

![error](https://user-images.githubusercontent.com/26309051/88404807-4bf83d80-cdc6-11ea-881e-90b1d2e1bddf.png)

My companies O365 uses ADFS.  Any ides what I can do as a workaround?
I have tried a subfilter, but can't seem to get it to work.

@0mikron, did you get something working?  Looks like you had the same issue.

Cheers

rico112 commented:
> > That’ll mean it’s not an office 365 account, it’s more like a Hotmail or Live account.
> > […](#)
> > On 17 Feb 2020, at 21:51, YusufHegazy _**@**_.***> wrote: ﻿ Hi! I'm trying to use this phishlet but on typing my email it redirects me to a https://login.live.com/ page displaying "invalid_request: The provided value for the input parameter 'redirect_uri' is not valid. The expected value is a URI which matches a redirect URI registered for this client application." does this have something to do with the phishlet, or it's on my side? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> is there any hotmail or live account yaml available ?

I also have that problem. redirect_uri' is not valid
Why is that?

cauan commented:
Hi @JamesCullum 

I had the same issue http......:443

I've created a sub_filter to fix it:
`  - {triggers_on: 'adfs.phishingdomain.com', orig_sub: 'login', domain: 'microsoftonline.com', search: 'adfs.phishingdomain.com/adfs/services/trust', replace: 'adfs.phishingdomain.com.au/adfs/services/trust', mimes: ['text/html', 'application/json', 'application/javascript']}`

But I found out that the sub_filters created by evilgix2 are executed after the sub_filters created by users, making my filter useless.

So I've this dirt sub_filter:

`  - {triggers_on: 'adfs.phishingdomain.com', orig_sub: 'login', domain: 'microsoftonline.com', search: 'adfs.phishingdomain.com/adfs/services/trust', replace: 'adfs.phishingdomain&#46;com.au/adfs/services/trust', mimes: ['text/html', 'application/json', 'application/javascript']}`

By replacing the . per its HTML encoded representation (&#46;), it is not matched by the filters automatically created by evilginx2.

It worked for me, it may help others.

donjuanme commented:
nice, thanks for that @cauan

Blussertje commented:
Hi everyone, 

We have been trying to setup a Phishing for ADFS using the o365.yaml template. A user is able to enter his/her username and is then redirected to adfs.phishing.com where he/she should be able to enter his password. However, when the user is redirected to the adfs.phishing.com domain, the browser gives the following error message: 
**Secure Connection Failed**

In the EvilGinx console, we receive the following error message:
**WARN: Cannot read TLS response from mitm'd server local error: tls: no renegotiation**

The adfs.phishing.com/ endpoint itself is accessible (error 404) and the certificate is trusted. 
We are not using any other proxies. 

Does anyone know what we are missing here? 

Thanks in advance!
Best regards.
